### ISSUE TITLE
Profiling GPU memory and compute for forward and backward

### ISSUE THREAD

**Issue by mseeger** [awslabs/keys_values#21]
**Is your feature request related to a problem? Please describe.**

At the moment, the time-memory trade-off for many choices (checkpointing, quantization, choice of cell size) is unknown. We also do not know where the most time is spent.

**Describe the solution you'd like**

Get a clear understand where the most time is spent in (1) inference and (2) gradient computation:

* What are improvements we can do without kernel implementation?
* How do arguments affect the dominating time/memory sources?

Use: https://docs.nvidia.com/nsight-compute/ProfilingGuide/#range-replay

**Comment by mseeger** [comment]
Start with profiling single forward pass (part of gradient computation, so with activation checkpointing).

```
CUDA_VISIBLE_DEVICES="0" PYTORCH_ALLOC_CONF=expandable_segments:True KEYSVALS_LOG_DIR="/home/ubuntu/out/finetune/ml_ws/lora/qwen3_4b/profile/lora_off/logs" nsys profile --trace=cuda,nvtx --output report_lora_off_1 python3 keys_values/__main__.py finetune_offload_lora Qwen/Qwen3-4B-Instruct-2507 --out_dir /home/ubuntu/out/finetune/ml_ws/lora/qwen3_4b/profile/lora_off --data LongBenchV2 --data.max_seq_length 150000 --data.metadata_dir /home/ubuntu/out/finetune/data --precision bf16-true --kv_cache.cache_length 32768 --kv_cache.chunk_size 2048 --kv_cache.verbose some --grad.layers_per_cell 1 --train.save_interval 10 --train.micro_batch_size 2 --train.global_batch_size 2 --eval.interval 10 --head_model seq_classification_on_logits --data.trainloader_longest_first True --attention_forward_temp_size_gb 2 --kv_cache.name h2o-torch-quantized8 --kv_cache.cpu_offload True --eval.initial_validation False --profile_grad_times 10

chunk_sizes     = [32768, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 199]
chunks_per_cell = [1, 12, 12, 12, 12, 10]
start's         = [0, 32768, 57344, 81920, 106496, 131072]

[DONE] -> report_lora_off_1.nsys-rep
```

Compare to:
```
CUDA_VISIBLE_DEVICES="0" PYTORCH_ALLOC_CONF=expandable_segments:True KEYSVALS_LOG_DIR="/home/ubuntu/out/finetune/ml_ws/lora/qwen3_4b/profile/lora_off_1_5/logs" nsys profile --trace=cuda,nvtx --output report_lora_off_1_5 python3 keys_values/__main__.py finetune_offload_lora Qwen/Qwen3-4B-Instruct-2507 --out_dir /home/ubuntu/out/finetune/ml_ws/lora/qwen3_4b/profile/lora_off_1_5 --data LongBenchV2 --data.max_seq_length 150000 --data.metadata_dir /home/ubuntu/out/finetune/data --precision bf16-true --kv_cache.cache_length 32768 --kv_cache.chunk_size 2048 --kv_cache.verbose some --grad.layers_per_cell 1 --train.save_interval 10 --train.micro_batch_size 2 --train.global_batch_size 2 --eval.interval 10 --head_model seq_classification_on_logits --data.trainloader_longest_first True --attention_forward_temp_size_gb 2 --kv_cache.name h2o-torch-quantized8 --kv_cache.cpu_offload True --eval.initial_validation False --profile_grad_times 10 --grad.chunks_per_cell_multiplier 1.5

chunk_sizes     = [32768, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 2048, 199]
chunks_per_cell = [4, 19, 19, 17]
start's         = [0, 38912, 77824, 116736]

[DONE] -> report_lora_off_1_5.nsys-rep
```

Use UI tool and look at processing of one cell:
```
1: 12 chunks: [81920 -- 106496]
- Block 0: 133.596 (0.59056)
  - layer_input_0:  92.7 (async_cp 78.5)
  - forward_0:     590.6
  - layer_input_1: 567.8 (async_cp 552.3)
  - forward_1:     597.6 (some small async_cp)
  - layer_input_2: 578.4 (async_cp 550.5)
  - forward_2:     598.4 (some small async_cp)
  - layer_input_3: 578.1 (async_cp 550.1)
  - forward_3:     592.4 (some small async_cp)

1.5: 19 chunks: [77824 -- 116736]
- Block 0: 124.585 (1.209)
  - layer_input_0: 147.2 (async_cp 122.5)
  - forward_0:     1209
  - layer_input_1: 566.9 (async_cp 545.3)
  - forward_1:     1219 (some small async_cp)
  - layer_input_2: 589.1 (async_cp 544.5)
  - forward_2:     1219 (some small async_cp)
  - layer_input_3: 589.6 (async_cp 545.1)
  - forward_3:     1220 (some small async_cp)
```

Questions/findings:

- 19/12 = 1.6, but 1220/600 = 2.0:
  Work on larger chunks should be faster.
  Something to do with eager kernel maybe? Compare against lastrec!
- Why `layer_input_X` not concurrent with `forward_X`?
  Maybe because of quantization?
- CPU offloading of KV cache may not be an issue?

**Comment by mseeger** [comment]
Hypotheses for these 250 ms to 290 ms `cudaMemcpyAsync` blocks:

- Layer input from CPU
- Head gradient from CPU

Try to put brackets around both of them!
